samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -18,47 +18,178 @@
+  private final Config config;
+  private final long taskShutdownMs;
+  private ExecutorService executorService;
+  private volatile SamzaContainer container = null;
+
+  @VisibleForTesting
+  JobCoordinatorListener jobCoordinatorListener = null;
+
+  // Latch used to synchronize between the JobCoordinator thread and the container thread, when the container is
+  // stopped due to re-balancing
+  private volatile CountDownLatch jcContainerLatch = new CountDownLatch(1);
+  private volatile boolean processorOnStartCalled = false;
+
+  SamzaContainer createSamzaContainer(ContainerModel containerModel, int maxChangelogStreamPartitions, JmxServer jmxServer) {
+    return SamzaContainer.apply(
+        containerModel,
+        config,
+        maxChangelogStreamPartitions,
+        jmxServer,
+        Util.<String, MetricsReporter>javaMapAsScalaMap(customMetricsReporter),
+        taskFactory);
+  }
+
+  JobCoordinatorListener createJobCoordinatorListener() {
+    return new JobCoordinatorListener() {
+
+      @Override
+      public void onJobModelExpired() {
+        if (container != null && container.getStatus().equals(SamzaContainerStatus.RUNNING)) {
+          container.shutdown(true);
+          boolean shutdownComplete = false;
+          try {
+            shutdownComplete = jcContainerLatch.await(taskShutdownMs, TimeUnit.MILLISECONDS);
+          } catch (InterruptedException e) {
+            LOGGER.warn("Container shutdown was interrupted!" + container.toString());
+          }
+          if (!shutdownComplete) {
+            LOGGER.warn("Container " + container.toString() + " may not have shutdown successfully. Stopping the processor.");
+            container = null;
+            stop();
+          } else {
+            LOGGER.info("Container " + container.toString() + " shutdown successfully");
+          }
+        }
+      }
+
+      @Override
+      public void onNewJobModel(String processorId, JobModel jobModel) {
+        if (!jobModel.getContainers().containsKey(processorId)) {
+          LOGGER.warn("JobModel did not contain the processorId. Stopping the processor.");
+          stop();
+        } else {
+          jcContainerLatch = new CountDownLatch(1);
+
+          SamzaContainerListener containerListener = new SamzaContainerListener() {
+            @Override
+            public void onContainerStart() {
+              if (!processorOnStartCalled) {
+                // processorListener is called on start only the first time the container starts.
+                // It is not called after every re-balance of partitions among the processors
+                processorOnStartCalled = true;
+                if (processorListener != null) {
+                  processorListener.onStart();
+                }
+              }
+            }
+
+            @Override
+            public void onContainerStop(boolean pauseByJm) {
+              if (jcContainerLatch != null) {
```

Comment:
Yeah. That should be fine **I think** Will feel more comfortable when we have more tests !